### PR TITLE
add checklist feature

### PR DESCRIPTION
**Pull request by lkisme**
add checklist feature, user can add multiple checklists, in which there can be multiple items, to a card

---

This change is [Reviewable](https://reviewable.io/reviews/wekan/wekan/630)

### CODE REVIEW

**Comment by lkisme** [comment]
Card with checklists
all checklist are checked

File: run-email.sh [hunk 1]
```diff
@@ -0,0 +1 @@
+MONGO_URL=mongodb://localhost:27017/wekan MAIL_URL=smtp://noreplytest:noreply110@smtp.126.com:25/ MAIL_FROM=noreplytest@126.com meteor
```
Comment (AlexanderS) [inline run-email.sh]: I think this shouldn't be here... And I think you should change your password **now**.

File: run.sh [hunk 2]
```diff
@@ -0,0 +1 @@
+MONGO_URL=mongodb://localhost:27017/wekan meteor
```
Comment (AlexanderS) [inline run.sh]: This shouldn't be here...

File: run-email.sh (hunk 1 above)
Comment (lkisme) [inline run-email.sh]: That's just an app password, no harm will be done, and it has been changed. I'll remove these 2 files from the diff

**Comment by brylie** [comment]
Great work!

**Comment by chuanqitan** [comment]
Great! Thanks for your work!

**Comment by brylie** [comment]
@mquandalle any timeline for PR review here?

**Comment by irfanitw** [comment]
Awesome work @lkisme this looks great!!!!!!!!

How do we get this pushed thru?

**Comment by smarek** [comment]
@lkisme great work on this feature, we really do appreciate it!

Few more things are left to be implemented, if anybody is willing to participate, see updated checklist here: https://github.com/wekan/wekan/issues/597#issue-153674576

**Comment by lkisme** [comment]
@mquandalle any schedule for reviewing and finalising the diff to be merged? so we can proceed on working the the feature further.

**Comment by brylie** [comment]
Given the lack of response/follow-up from @mquandalle, it might be reasonable for the Wekan community to discuss a community-maintained fork. I will open an issue for discussion.

**Comment by brylie** [comment]
See issue #640 for a discussion of a new Wekan community governance structure.

**Comment by mquandalle** [comment]
Reviewed 15 of 17 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8emKnOR4KAvIudxEo:-KN8emKoT4uRyOzjAiiP:877119056) (no related file):*
Thank you for the pull request and please excuse me for the delay in reviewing it. I’ve left some inline comments on specifics code issues that could easily be fixed and that would polish the feature behavior and its integration in the Wekan code base.

As a broader open question I wonder if we really need/want to introduce a checklistItem entity in the model, as its schema looks very similar to the card schema with small semantics changes (a card is “archived” whereas a checklistItem is “finished”). What if instead in order to keep the number of concepts low, we define that a checklistItem is actually a card whose parent happen to be another card; we would just have to change the `listId` property of the card model to an abstracted `parentId`. What do you think of this idea?

---

*[client/components/cards/checklist.jade, line 1 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8ZfcinH6eFuyLKN8q:-KN8ZfcinH6eFuyLKN8r:756379843) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.jade#L1)):*
> ```Jade
> template(name="checklists")
> ```

I would rename these three files `checklists.{jade,style.js}` to be consistent with the rest of the code base.

---

*[client/components/cards/checklist.js, line 4 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8_pfsaxDhuBBrbAG9:-KN8_pftAYEF6GvdBXOV:-191290257) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.js#L4)):*
> ```JavaScript
>   addChecklist(event) {
>     event.preventDefault();
>     const textarea = $(event.currentTarget).find('textarea');
> ```

We try to avoid jQuery where it’s not necessary. In this case you can use `this.find` instead of `$(event.currentTarget).find`. Also please give a `js` prefixed CSS class to the textearea field, like `.js-checklist` or something. Also you don’t need jQuery either to `.trim()` the string as it is now a standard ES6 method.

---

*[client/components/cards/checklist.js, line 10 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8b2lX-JBW34DNeU1n:-KN8b2lYsZYWP8iwMcfp:2114971752) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.js#L10)):*
> ```JavaScript
>       cardId,
>       title,
>     });
> ```

After the checklist is created, you should open the `addChecklistItem` form, so that the user could continue to type its item without leaving the keyboard.

---

*[client/components/cards/checklist.js, line 17 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8bCziQ8RZqUn3V4kc:-KN8bCziQ8RZqUn3V4kd:-483080134) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.js#L17)):*
> ```JavaScript
>     const title = textarea.val().trim();
>     const checklist = this.currentData().checklist;
>     checklist.addItem(title);
> ```

Same thing here, after a checklist item is added, you should leave the form open (as we do to add a card) so that the user could write further items. You can also leave a JS comment saying that the logic is pretty similar to the “addCard” logic and could probably be abstracted away.

---

*[client/components/cards/checklist.js, line 24 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8_05SovL444gU3yV1:-KN8_05SovL444gU3yV2:-1942202077) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.js#L24)):*
> ```JavaScript
>     const title = textarea.val().trim();
>     const checklistId = this.currentData().checklist._id;
>     Checklists.update(checklistId, {
> ```

Please define a mutation in `models/checklists.js` and call it from here. We centralize all defined mutations in the model definition.

---

*[client/components/cards/checklist.js, line 53 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8bdId4L-1hcgeNyJL:-KN8bdIeyFpf0jiPylO8:-212899952) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.js#L53)):*
> ```JavaScript
>   },
>   pressKey(event) {
>     if (event.keyCode === 13) {
> ```

Leave a comment explaining the intent here.

---

*[client/components/cards/checklist.styl, line 1 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8dHRQrR7_tfQwwFiu:-KN8dHRQrR7_tfQwwFiv:422985833) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.styl#L1)):*
> ```Stylus
> .js-add-checklist
> ```

Just to have a consistent formating, could you please: (1) add colons after the property name and (2) and a new line before selectors.

---

*[client/components/cards/checklist.styl, line 62 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8dBNFh0ZkwyLpudpe:-KN8dBNFh0ZkwyLpudpf:317069916) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.styl#L62)):*
> ```Stylus
>     display inline-block
> 
>     
> ```

Remove the empty lines at the end of this file

---

*[client/components/cards/minicard.styl, line 107 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8bqgd-DdZjLG4uZ71:-KN8bqgd-DdZjLG4uZ72:1636830067) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/minicard.styl#L107)):*
> ```Stylus
>       .badge-icon,
>       .badge-text
>         vertical-align: middle//didn't figure why use top, it'd be easier to fill bg if it's middle. This was introduced in commit "91cfcf7b12b5e7c137c2e765b2c378dde6b82966" & "* Improve the design of the minicards badges" was mentioned.
> ```

I think it was for the comments badge icon. I’m ok with middle, but the comment icon should probably be moved a few pixels up.

---

*[i18n/en.i18n.json, line 45 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8csbZmnvqWmvsw_UP:-KN8csbZmnvqWmvsw_UQ:-828173271) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/i18n/en.i18n.json#L45)):*
> ```JSON
>     "add-card": "Add a card",
>     "add-checklist": "Add a checklist",
>     "add-checklist-item": "Add an item to checklist",
> ```

If the checklist as a name, we should reference it in the activity feed.

---

*[models/checklists.js, line 33 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8e947SMl3ArD2IltI:-KN8e948j4bnZQspMIBF:1118040180) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/models/checklists.js#L33)):*
> ```JavaScript
>   itemCount () {
>     if (!this.items) {
>       this.items = [];
> ```

You could set a defaultValue in the schema above.

---

*[models/checklists.js, line 144 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8eNZBGAR94FNGMRpf:-KN8eNZBGAR94FNGMRpg:-1870802478) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/models/checklists.js#L144)):*
> ```JavaScript
>   });
> 
>   //TODO: so there will be no activity for adding item into checklist, maybe will be implemented in the future.
> ```

Yes, I’m thinking of a refactoring that would allow tracking all mutations in the activity feed automatically.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/wekan/wekan/630)*

**Comment by lkisme** [comment]
Review status: 9 of 15 files reviewed at latest revision, 15 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8emKnOR4KAvIudxEo:-KNHS2_juQlSyeC7c0To:203922120) (no related file):*
<details><summary><i>Previously, mquandalle (Maxime Quandalle) wrote…</i></summary>
> Thank you for the pull request and please excuse me for the delay in reviewing it. I’ve left some inline comments on specifics code issues that could easily be fixed and that would polish the feature behavior and its integration in the Wekan code base.
> 
> As a broader open question I wonder if we really need/want to introduce a checklistItem entity in the model, as its schema looks very similar to the card schema with small semantics changes (a card is “archived” whereas a checklistItem is “finished”). What if instead in order to keep the number of concepts low, we define that a checklistItem is actually a card whose parent happen to be another card; we would just have to change the `listId` property of the card model to an abstracted `parentId`. What do you think of this idea?
</details>
Yes, you are absolutely right about the idea of keeping number of concepts low, but this will introduce several other problems, 1st there will be hierarchy like a card(checklist item) belongs to a checklist that belongs to a card, 2nd a card usually is seen as a user story while a checklist item should be seen as a sub task(relatively smaller), 3rd to designe a abstract parent would introduce more complexities, how do you think about these?

---

*[client/components/cards/checklist.jade, line 1 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8ZfcinH6eFuyLKN8q:-KNIDGlV3lzV6iqSvbbk:-499146009) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.jade#L1)):*
<details><summary><i>Previously, mquandalle (Maxime Quandalle) wrote…</i></summary>
> I would rename these three files `checklists.{jade,style.js}` to be consistent with the rest of the code base.
</details>
Done.

---

*[client/components/cards/checklist.js, line 4 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8_pfsaxDhuBBrbAG9:-KNIDcv-1uuJP_spqqMe:-1878343788) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.js#L4)):*
<details><summary><i>Previously, mquandalle (Maxime Quandalle) wrote…</i></summary>
> We try to avoid jQuery where it’s not necessary. In this case you can use `this.find` instead of `$(event.currentTarget).find`. Also please give a `js` prefixed CSS class to the textearea field, like `.js-checklist` or something. Also you don’t need jQuery either to `.trim()` the string as it is now a standard ES6 method.
</details>
Do you mean add those classes into template file, or use the class to find the element in JS? cause there are already classes for each of them. like "textarea.js-add-checklist-item"

---

*[client/components/cards/checklist.js, line 10 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8b2lX-JBW34DNeU1n:-KNLCbg513q2YNbH3Z6r:631951021) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.js#L10)):*
<details><summary><i>Previously, mquandalle (Maxime Quandalle) wrote…</i></summary>
> After the checklist is created, you should open the `addChecklistItem` form, so that the user could continue to type its item without leaving the keyboard.
</details>
Yes, that would be better, although I couldn't find a proper way to implement that, might need some help with the implementation.

---

*[client/components/cards/checklist.js, line 17 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8bCziQ8RZqUn3V4kc:-KNLDJ54vlfTEys_-T_V:487508482) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.js#L17)):*
<details><summary><i>Previously, mquandalle (Maxime Quandalle) wrote…</i></summary>
> Same thing here, after a checklist item is added, you should leave the form open (as we do to add a card) so that the user could write further items. You can also leave a JS comment saying that the logic is pretty similar to the “addCard” logic and could probably be abstracted away.
</details>
same as above, it would be great if there is a abstracted way to implement this logic.

---

*[client/components/cards/checklist.js, line 24 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8_05SovL444gU3yV1:-KNIDN7uapghVnj01qNx:-499146009) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.js#L24)):*
<details><summary><i>Previously, mquandalle (Maxime Quandalle) wrote…</i></summary>
> Please define a mutation in `models/checklists.js` and call it from here. We centralize all defined mutations in the model definition.
</details>
Done.

---

*[client/components/cards/checklist.js, line 53 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8bdId4L-1hcgeNyJL:-KNIEJm_YWzuaEurZszk:-499146009) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.js#L53)):*
<details><summary><i>Previously, mquandalle (Maxime Quandalle) wrote…</i></summary>
> Leave a comment explaining the intent here.
</details>
Done.

---

*[client/components/cards/checklist.styl, line 1 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8dHRQrR7_tfQwwFiu:-KNIKwKnh65sHTarGxXU:-499146009) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.styl#L1)):*
<details><summary><i>Previously, mquandalle (Maxime Quandalle) wrote…</i></summary>
> Just to have a consistent formating, could you please: (1) add colons after the property name and (2) and a new line before selectors.
</details>
Done.

---

*[client/components/cards/checklist.styl, line 62 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8dBNFh0ZkwyLpudpe:-KNIKu5U29t-K8uc33aG:-499146009) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/checklist.styl#L62)):*
<details><summary><i>Previously, mquandalle (Maxime Quandalle) wrote…</i></summary>
> Remove the empty lines at the end of this file
</details>
Done.

---

*[client/components/cards/minicard.styl, line 107 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8bqgd-DdZjLG4uZ71:-KNIOnocAkquvmbAT5Ue:-499146009) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/client/components/cards/minicard.styl#L107)):*
<details><summary><i>Previously, mquandalle (Maxime Quandalle) wrote…</i></summary>
> I think it was for the comments badge icon. I’m ok with middle, but the comment icon should probably be moved a few pixels up.
</details>
Done.

---

*[i18n/en.i18n.json, line 45 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8csbZmnvqWmvsw_UP:-KNIRJvZj5DGZYBDzOve:684751354) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/i18n/en.i18n.json#L45)):*
<details><summary><i>Previously, mquandalle (Maxime Quandalle) wrote…</i></summary>
> If the checklist as a name, we should reference it in the activity feed.
</details>
didn't get it.

---

*[models/checklists.js, line 33 \[r2\]](https://reviewable.io:443/reviews/wekan/wekan/630#-KN8e947SMl3ArD2IltI:-KNIQwaYOIRDSb3VUcMX:-499146009) ([raw file](https://github.com/wekan/wekan/blob/4d7560069b57d4cb95816d6e10af1d98cf003f15/models/checklists.js#L33)):*
<details><summary><i>Previously, mquandalle (Maxime Quandalle) wrote…</i></summary>
> You could set a defaultValue in the schema above.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/wekan/wekan/630)*

**Comment by nata-g** [comment]
@mquandalle Any ETA when this feature would be merged? Seems to be one of the most requested feature :-)

**Comment by darkblinded** [comment]
Any updates on this? I had the feeling, this was one of the most requested features.

**Comment by rezagarrow** [comment]
Also very interested in this.  The only thing keeping our team from moving off Trello.

**Comment by lkisme** [comment]
@mquandalle It's been a while since your last comment on this, still looking forward to your response

**Comment by brylie** [comment]
@lkisme we ended up forking the project, primarily based on the lack of response to this pull request.

Please open the pull request against the [Wefork/Wekan project repository](https://github.com/wefork/wekan), and mention that it closes [issue #298](https://github.com/wekan/wekan/issues/298).

**Comment by whittssg** [comment]
Did anyone try merging this on a local copy? I tried and it looks great but i can get the checkboxes to actually check.

**Comment by lkisme** [comment]
@whittssg Does other checkboxes work in your local copy? the implementation of the checkbox in this PR is same with any other checkboxes in this project, you can try to run Wekan without this PR, then merge it into your copy. hope this will help.

**Comment by whittssg** [comment]
@lkisme I go this working yesterday, i just forgot to update my message. (i think i talked to you in the gitter channel). Thanks

**Comment by lkisme** [comment]
@whittssg NP, glad to see this feature works

**Comment by ocdtrekkie** [comment]
FYI for everyone following this PR: This has been merged. So, update on that.

**Pull request closed without merging**